File: tenstorrent/course-sched-2.cpp
#include <unordered_map>
#include <vector>
#include <iostream>
#include <queue>

// Problem : Given a list of prequistes [0, 1] we need to output a final order or courses take to determine compatbility and chronology
// Example 1: [[0,1], [1,0]]
// 0 -> 1 and 1 -> 0
// Output: -1
// 
// Example 2 : [[1,0], [2, 1], [3, 2]]
// 0 -> 1 -> 2 -> 3
// Output: [0,1,2,3]
// 
// Example 3: [[1,0], [2,1], [3,1], [4, 3]]
// 0 -> 1, 1 -> 2, 1 -> 3, 3 -> 4
// Output: [0, 1, 2, 3, 4]

class Solution {
    public:
    std::vector<int> solve(std::vector<std::pair<int, int>> prereq, int numCourses){
        // Approach: create a adjency list of all node and their outgoing neihbor
        // create a indegree vector that will hold the indegree of each course. 
        std::unordered_map<int, std::vector<int>> mp(numCourses);
        std::vector<int> indeg(numCourses, 0);
        
        for (auto p : prereq){
            // first -> course , second -> dependency
            mp[p.second].push_back(p.first);
            indeg[p.first]++;
        }
        
        // Example: (2) it is going to be 1 for most and 0 for the ones we are going to start processing with
        std::queue<int> q;
        for (int i = 0; i < numCourses; i++){
            int v = indeg[i];
            if (v == 0)  q.push(i);
        }
        
        // we can 0 indegrees to the queue, process their each nbr, decrease the indegree 
        // and push them back in the queue if their indegree becomes 0
        std::vector<int> res;
        while(!q.empty()){
            auto v = q.front();
            q.pop();
            res.push_back(v);
            
            for(int nbr : mp[v]){
               indeg[nbr]--; 
               if (indeg[nbr] == 0){
                   q.push(nbr);
               }
            }
        }
        
        return res;
    };
};

int main(){
    Solution sol;
    auto a = sol.solve({{1,0}, {2, 1}, {3, 2}}, 4);
    
    for (int i : a){
        std::cout << i << " -> ";
    }
    
    std::cout << std::endl;
    return 0;
};

File: course_sched_2.py
# There are a total of numCourses courses you have to take, labeled from 0 to numCourses - 1. You are given an array prerequisites where prerequisites[i] = [ai, bi] indicates that you must take course bi first if you want to take course ai.
# 
# bi -> ai

# For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.

# Return the ordering of courses you should take to finish all courses. If there are many valid answers, return any of them. If it is impossible to finish all courses, return an empty array.

# Approach: Create a adj_list for each course: 
# which courses a single course depend on the 
# which course has no dependency has indegree of 0
# we also have to track the indegree number -> 
# we maintain a queue and push all the node wd indegree == 0
# and then check all their neighbors decrease their adjancey and if their indegree 
# then thay are safe to add
# 
from typing import Deque, List


class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = {i: [] for i in range(numCourses) }
        indeg = [0] * numCourses
        
        for c, d in prerequisites:
            adj[d].append(c) 
            indeg[c] += 1
            
        queue = Deque(i for i in range(numCourses) if indeg[i] == 0)
        res = []
        
        while queue:
           v = queue.popleft()
           res.append(v)
           for nbr in adj[v]:
               indeg[nbr] -= 1
               if indeg[nbr] == 0:
                   queue.append(nbr)
               
        return res

if __name__ == "__main__":
    print("Test 1:", Solution().findOrder(4, [[1,0],[2,0],[3,1],[3,2]]))
